### PR TITLE
chore(*): update election example

### PR DESCRIPTION
**Pull request by bbbmj**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:

/kind cleanup

**What this PR does / why we need it**:

Clean up and simplify the leader election example.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 1]
```diff
@@ -56,21 +55,17 @@ func buildConfig(kubeconfig string) (*rest.Config, error) {
 func main() {
 	klog.InitFlags(nil)
 
+	var leaderElect bool
 	var kubeconfig string
 	var leaseLockName string
 	var leaseLockNamespace string
-	var id string
 
+	flag.BoolVar(&leaderElect, "leader-elect", true, "whether enable leader elect")
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: Why do we need this for a leader election example?

File: staging/src/k8s.io/client-go/examples/leader-election/README.md [hunk 2]
```diff
@@ -4,18 +4,19 @@ This example demonstrates how to use the leader election package.
...
+go run main.go -kubeconfig=/my/config -logtostderr=true
 
-# second terminal 
-go run *.go -kubeconfig=/my/config -logtostderr=true -id=2
+# second terminal
+go run main.go -kubeconfig=/my/config -logtostderr=true
 
 # third terminal
-go run *.go -kubeconfig=/my/config -logtostderr=true -id=3
+go run main.go -kubeconfig=/my/config -logtostderr=true
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/examples/leader-election/README.md]: -leader-elect

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 3]
```diff
@@ -56,21 +55,17 @@ func buildConfig(kubeconfig string) (*rest.Config, error) {
...
 	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
-	flag.StringVar(&id, "id", "", "the holder identity name")
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: Leave this flag but change it to:

```
flag.StringVar(&id, "id", uuid.New().String(), "the holder identity name")
```

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 4]
```diff
@@ -110,10 +99,48 @@ func main() {
...
+		}(ctx)
+
+		run(ctx)
+
+		panic("unreachable")
+	}
+
+	klog.Infof("Run with leader election")
+
+	id, err := os.Hostname()
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: We always expect the hostname to be the same in this example since it'll be run from the same machine.

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2019 The Kubernetes Authors.
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: i don't think the year should be updated here

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 6]
```diff
@@ -124,38 +121,29 @@ func main() {
 		// get elected before your background loop finished, violating
 		// the stated goal of the lease.
 		ReleaseOnCancel: true,
-		LeaseDuration:   60 * time.Second,
-		RenewDeadline:   15 * time.Second,
-		RetryPeriod:     5 * time.Second,
+		LeaseDuration:   15 * time.Second,
+		RenewDeadline:   10 * time.Second,
+		RetryPeriod:     2 * time.Second,
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: what is the reason for these changes?

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 7]
```diff
@@ -124,38 +121,29 @@ func main() {
...
 				// usually put your code
-				klog.Infof("%s: leading", id)
+				run(ctx)
 			},
 			OnStoppedLeading: func() {
-				// we can do cleanup here, or after the RunOrDie method
-				// returns
-				klog.Infof("%s: lost", id)
+				// we can do cleanup here
+				klog.Fatalf("leader lost: %s", id)
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: why fatal here?

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 8]
```diff
@@ -124,38 +121,29 @@ func main() {
...
+				klog.Infof("new leader elected: %s", identity)
 			},
 		},
 	})
 
-	// because the context is closed, the client should report errors
-	_, err = client.CoordinationV1().Leases(leaseLockNamespace).Get(leaseLockName, metav1.GetOptions{})
-	if err == nil || !strings.Contains(err.Error(), "the leader is shutting down") {
-		log.Fatalf("%s: expected to get an error when trying to make a client call: %v", id, err)
-	}
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: what is the reason for removing that from the example?

**Review by yliaog — COMMENTED** [review]
please fill in the PR description, so it's easier for people to understand what it does without reading the code.

File: staging/src/k8s.io/client-go/examples/leader-election/main.go (hunk 6 above)
Comment (bbbmj) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: From kube-controller-manager, and we can adjust the value according to personal consideration.

File: staging/src/k8s.io/client-go/examples/leader-election/main.go (hunk 8 above)
Comment (bbbmj) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: I think it's ilelegant as we need to confirm error with `strings.Contains()`,  and remove that can simplify the example.

File: staging/src/k8s.io/client-go/examples/leader-election/main.go (hunk 7 above)
Comment (bbbmj) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: only one place to put clean up code is better for reading.

File: staging/src/k8s.io/client-go/examples/leader-election/README.md [hunk 9]
```diff
@@ -4,18 +4,19 @@ This example demonstrates how to use the leader election package.
 
 ## Running
 
-Run the following three commands in separate terminals. Each terminal needs a unique `id`.
+Run the following three commands in separate terminals.
```
Comment (bbbmj) [inline staging/src/k8s.io/client-go/examples/leader-election/README.md]:
```suggestion
Run the following three commands in separate terminals. Each terminal needs a unique `id`.
```

**Comment by bbbmj** [comment]
Just a small simplification.

**Comment by bbbmj** [comment]
cc @mikedanese @liggitt 

ok to merge? and I will make a new PR to complete healthz later

File: staging/src/k8s.io/client-go/examples/leader-election/main.go (hunk 6 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: as an example, a one minute lease seems ok

File: staging/src/k8s.io/client-go/examples/leader-election/main.go (hunk 7 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: exiting when leader lease is lost is reasonable, but this exits with a non-zero return code, which we probably don't want

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 10]
```diff
@@ -124,38 +121,29 @@ func main() {
...
-
-	// we no longer hold the lease, so perform any cleanup and then
-	// exit
-	log.Printf("%s: done", id)
+	panic("unreachable")
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: I would drop the panic

File: staging/src/k8s.io/client-go/examples/leader-election/main.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: is it good to point people in the direction of automatic random identities for leader-elected controllers, or should we leave it requiring an id provided via invocation?

Comment (liggitt) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: I guess this lets them run n replicas with a single invocation

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 11]
```diff
@@ -62,15 +59,11 @@ func main() {
 	var id string
 
 	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
-	flag.StringVar(&id, "id", "", "the holder identity name")
+	flag.StringVar(&id, "id", uuid.New().String(), "the holder identity name")
 	flag.StringVar(&leaseLockName, "lease-lock-name", "example", "the lease lock resource name")
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: I would drop the default lease lock name and namespace and require values be passed in via args. Assuming people will copy this verbatim, we don't want lots of random controllers competing on the example/default name/namespace

File: staging/src/k8s.io/client-go/examples/leader-election/main.go (hunk 10 above)
Comment (bbbmj) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: done

File: staging/src/k8s.io/client-go/examples/leader-election/main.go (hunk 11 above)
Comment (bbbmj) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: done

**Comment by bbbmj** [comment]
@liggitt PTAL

File: staging/src/k8s.io/client-go/examples/leader-election/main.go [hunk 12]
```diff
@@ -131,31 +135,21 @@ func main() {
...
 				// we're notified when we start - this is where you would
 				// usually put your code
-				klog.Infof("%s: leading", id)
+				run(ctx)
 			},
 			OnStoppedLeading: func() {
 				// we can do cleanup here, or after the RunOrDie method
 				// returns
-				klog.Infof("%s: lost", id)
+				klog.Infof("leader lost: %s", id)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/examples/leader-election/main.go]: does this still result in the process exiting? if not, I'd take `os.Exit(0)` here

Comment (bbbmj) [reply staging/src/k8s.io/client-go/examples/leader-election/main.go]: done

File: staging/src/k8s.io/client-go/examples/leader-election/README.md [hunk 13]
```diff
@@ -7,15 +7,16 @@ This example demonstrates how to use the leader election package.
 Run the following three commands in separate terminals. Each terminal needs a unique `id`.
 
 ```bash
-# first terminal 
-go run *.go -kubeconfig=/my/config -logtostderr=true -id=1
+# first terminal
+go run main.go -kubeconfig=/path/to/kebuconfig -logtostderr=true -lease-lock-name=example -lease-lock-namespace=default -id=1
```
[the remark this answers is not in the document]
Comment (bbbmj) [reply staging/src/k8s.io/client-go/examples/leader-election/README.md]: > typo, should be kubeconfig (applies to all examples)

fixed.

> do the flag names require -- instead of -?

both `-` and `--` is ok to use. we can get more information from https://golang.org/pkg/flag/#hdr-Command_line_flag_syntax

**Comment by bbbmj** [comment]
@liggitt PTAL
